docs/api/web-contents.md

```diff
@@ -1081,6 +1114,17 @@ Works like `executeJavaScript` but evaluates `scripts` in an isolated context.
 
 Ignore application menu shortcuts while this web contents is focused.
 
+#### `contents.setWindowOpenOverride(handler)`
+
+* `handler` Function<Boolean | BrowserWindowConstructorOptions>
+  * `details` Object
+    * `url` String - The _resolved_ version of the URL passed to `window.open()`. e.g. opening a window with `window.open('foo')` will yield something like `https://the-origin/the/current/path/foo`.
+    * `frameName` String - Name of the window provided in `window.open()`
+  Returns `Boolean | BrowserWindowConstructorOptions` - `false` cancels the creation of the new window. `true` results in the default behavior: a window is created. Returning `BrowserWindowConstructorOptions` allows customization of the created window. Returning `undefined` will result in a console error and cancellation of the window.
```

Comment:
I'd argue that the API you've proposed is also odd, and features a magical property with unclear semantics that lack precedent. From the function signature, it's unclear that it's even possible to override the options. It makes use of parameter mutation, which is [widely considered to be bad practice](https://eslint.org/docs/rules/no-param-reassign.html). Whereas in what I've proposed, the semantics are clear and idiomatic: the handler will be passed inputs and return outputs, and those outputs will affect the course of the program. I'd hardly call it magical that the program's response changes given a function's varying return values.

At first blush, it does seem that event semantics would be appropriate, but _notification_ is not the primary goal of this API, _consultation_ is. And there's no real precedent for consultation semantics in javascript that I can think of, because normally it's possible to get consultation with the union of notification (`event.preventDefault()`) and action (`window.doTheThingMyWay()`). But in this case, due to the precariousness of where we've had to hook into Chrome, the program has to take the action, rather than being allowed to delegate to the consumer. I agree that it's not an ideal API, but as so often happens in Electron, the ideal API is technically infeasible.

Reply:
I don't think `event.foo = bar` lacks precedent; we do the same thing in sync IPC with `event.returnValue` ([docs](https://www.electronjs.org/docs/api/ipc-main#sending-messages)). FWIW, that API is also present in the web, for instance in the [beforeunload event](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event). I'll grant it's unusual, though.

With the return-style API, is there any way to inspect what the default options would be? How would you express e.g. "do the default, but set the x/y coordinates"?

Reply:
Yes, the way to represent it would just be

```
webContents.setWindowOpenOverride(() => ({
  x: 500,
  y: 500,
}))
```

> is there any way to inspect what the default options would be?

It would be the same as creating a new BrowserWindow insofar as there are defaults, but they are inscrutable to the consumer. This is a departure from the way things work now, as I mention in the PR description:

> Non-security related BrowserWindowConstructorOptions are no longer inherited from the parent. This was confusing when you had to "undo" preferences that may have been set in a different part of your app.